man/man8/gentrace.8
.TH gentrace 8  "2016-02-11" "USER COMMANDS"
.SH NAME
gentrace \- Trace a function and display a histogram or summary of its parameter values. Uses Linux eBPF/bcc.
.SH SYNOPSIS
.B gentrace [-h] [-p PID] [-z STRING_SIZE] [-i INTERVAL] [-c COUNT] specifier [specifier ...]
.SH DESCRIPTION
gentrace attaches to function entry and exit points, collects specified parameter
values, and stores them in a histogram or a raw counting collection that counts
the number of times a parameter value occurred. It can also filter parameter
values and instrument multiple entry points at once.

This currently only works on x86_64. Check for future versions.
.SH REQUIREMENTS
CONFIG_BPF and bcc.
.SH OPTIONS
.TP
\-h
Print usage message.
.TP
\-p PID
Trace only functions in the process PID. This filter will only apply to user-space
functions. If you only provide kernel-space probes, the filter is ignored.
.TP
\-z STRING_SIZE
When collecting string arguments (of type char*), collect up to STRING_SIZE 
characters. Longer strings will be truncated.
.TP
-i INTERVAL
Print the collected data every INTERVAL seconds. The default is 1 second.
.TP
-c COUNT
Print the collected data COUNT times and then exit.
.TP
SPECIFIER
One or more probe specifications that instruct gentrace which functions to
probe, which parameters to collect, how to aggregate them, and whether to perform
any filtering. See SPECIFIER SYNTAX below.
.SH SPECIFIER SYNTAX
The general specifier syntax is as follows:

.B <raw|hist>[-ret]:[library]:function(signature)[:type:expr[:filter]]
.TP
Probe type \- "raw", "hist", "raw-ret", "hist-ret".
Indicates where to place the probe and whether the probe should collect raw
event information, or aggregate the collected values into a histogram. Raw 
probes will collect the number of times every parameter value was observed,
whereas histogram probes will collect the parameter values into a histogram.
Only integral types can be used with histogram probes; there is no such limitation
for raw probes. Suffix with \-ret to indicate that the probe should be placed
at function return. This probe can only use the pseudo-variable @retval, which
is an alias for the function's return value.
.TP
Library containing the probe.
Specify the full path to the .so or executable file where the function to probe
resides. Alternatively, you can specify just the lib name: for example, "c"
refers to libc. If no library name is specified, the kernel is assumed.
.TP
The function to probe, and its signature.
The function name must match exactly for the probe to be placed. The signature,
on the other hand, is only required if you plan to collect parameter values 
based on that signature. For example, if you only want to collect the first
parameter, you don't have to specify the rest of the parameters in the signature.
.TP
The type of the expression to capture.
This is the type of the keys in the histogram or raw event collection that are
collected by the probes.
.TP
The expression to capture.
These are the values that are assigned to the histogram or raw event collection.
You may use the parameters directly, or valid C expressions that involve the
parameters, such as "size % 10".
.TP
A filter applied to the captured data.
Only parameter values that pass the filter will be collected. This is any valid
C expression that refers to the parameter values, such as "fd == 1 && length > 16".
.SH EXAMPLES
.TP
Print a histogram of allocation sizes passed to kmalloc:
#
.B gentrace.py 'hist::__kmalloc(u64 size):u64:size'
.TP
Print a raw count of how many times process 1005 called malloc with an allocation size of 16 bytes:
#
.B gentrace.py -p 1005 'raw:c:malloc(size_t size):size_t:size:size==16'
.TP
Snoop on all strings returned by gets():
#
.B gentrace.py 'raw-ret:c:gets():char*:@retval'
.TP
Print raw counts of how many times writes were issued to a particular file descriptor number, in process 1005:
#
.B gentrace.py -p 1005 'raw:c:write(int fd):int:fd'
.TP
Print a histogram of error codes returned by read() in process 1005:
#
.B gentrace.py -p 1005 'hist-ret:c:read()'
.TP
Print a histogram of buffer sizes passed to write() across all processes, where the file descriptor was 1 (STDOUT):
#
.B gentrace.py 'hist:c:write(int fd, const void *buf, size_t count):size_t:count:fd==1'
.TP
Count fork() calls in libc across all processes:
#
.B gentrace.py 'raw:c:fork'
.TP
Print histograms of sleep() and nanosleep() parameter values:
#
.B gentrace.py 'hist:c:sleep(u32 seconds):u32:seconds' 'hist:c:nanosleep(struct timespec { time_t tv_sec; long tv_nsec; } *req):long:req->tv_nsec'
.TP
Spy on writes to STDOUT performed by process 2780, up to a string size of 120 characters:
#
.B gentrace.py -p 2780 -z 120 'raw:c:write(int fd, char* buf, size_t len):char*:buf:fd==1'
.SH SOURCE
This is from bcc.
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH AUTHOR
Sasha Goldshtein


debian/changelog
bcc (0.1.7-1) unstable; urgency=low

  * Tracing features and bugfixes
  * Built against LLVM 3.8 HEAD

 -- Brenden Blanco <bblanco@plumgrid.com>  Mon, 12 Oct 2015 16:47:09 +0000

bcc (0.1.6-1) unstable; urgency=low

  * Stability fixes
  * Improvements to python API
  * Tracing features
  * Support for kernel 4.2 features

 -- Brenden Blanco <bblanco@plumgrid.com>  Wed, 02 Sep 2015 16:23:19 +0000

bcc (0.1.5-1) unstable; urgency=low

  * Initial release

 -- Brenden Blanco <bblanco@plumgrid.com>  Mon, 06 Jul 2015 18:04:28 +0000


tools/gentrace.py
#!/usr/bin/env python
#
# gentrace.py   Trace a function and display a histogram or summary of its
#               parameter values. 
#
# USAGE: gentrace.py [-h] [-p PID] [-z STRING_SIZE] [-i INTERVAL]
#                    [-c COUNT] specifier [specifier ...]
#
# Licensed under the Apache License, Version 2.0 (the "License")
# Copyright (C) 2016 Sasha Goldshtein.

from bcc import BPF
from time import sleep, strftime
import argparse

class Specifier(object):
        text = """
DATA_DECL

int PROBENAME(struct pt_regs *ctx SIGNATURE)
{
        KEY_EXPR
        if (!(FILTER)) return 0;
        COLLECT
        return 0;
}
"""

        def __init__(self, specifier, pid):
                self.raw_spec = specifier 
                parts = specifier.strip().split(':')
                if len(parts) < 3 or len(parts) > 6:
                        raise ValueError("invalid specifier format")
                self.type = parts[0]    # hist or raw
                self.is_ret_probe = self.type.endswith("-ret")
                if self.is_ret_probe:
                        self.type = self.type[:-len("-ret")]
                if self.type != "hist" and self.type != "raw":
                        raise ValueError("unrecognized probe type")
                self.library = parts[1]
                fparts = parts[2].split('(')
                if len(fparts) != 2:
                        raise ValueError("invalid specifier format")
                self.function = fparts[0]
                self.signature = fparts[1][:-1]
                if len(parts) >= 5:
                        self.expr_type = parts[3]
                        self.expr = parts[4]
                else:
                        self.expr_type = \
                                "u64" if not self.is_ret_probe else "int"
                        self.expr = "1" if not self.is_ret_probe else "@retval"
                self.expr = self.expr.replace("@retval",
                                              "(%s)ctx->ax" % self.expr_type)
                self.filter = None if len(parts) != 6 else parts[5]
                self.pid = pid
                self.probe_func_name = self.function + "_probe"
                self.probe_hash_name = self.function + "_hash"

        def _is_string_probe(self):
                return self.expr_type == "char*" or self.expr_type == "char *"

        def generate_text(self, string_size):
                program = self.text.replace("PROBENAME", self.probe_func_name)
                signature = "" if len(self.signature) == 0 \
                               else "," + self.signature
                program = program.replace("SIGNATURE", signature)
                if self._is_string_probe():
                        decl = """
struct %s_key_t { char key[%d]; };
BPF_HASH(%s, struct %s_key_t, u64);
""" \
                        % (self.function, string_size,
                           self.probe_hash_name, self.function)
                        collect = "%s.increment(__key);" % self.probe_hash_name
                        key_expr = """
struct %s_key_t __key = {0};
bpf_probe_read(&__key.key, sizeof(__key.key), %s);
""" \
                        % (self.function, self.expr)
                elif self.type == "raw":
                        decl = "BPF_HASH(%s, %s, u64);" % \
                                (self.probe_hash_name, self.expr_type)
                        collect = "%s.increment(__key);" % self.probe_hash_name
                        key_expr = "%s __key = %s;" % \
                                   (self.expr_type, self.expr)
                elif self.type == "hist":
                        decl = "BPF_HISTOGRAM(%s, %s);" % \
                                (self.probe_hash_name, self.expr_type)
                        collect = "%s.increment(bpf_log2l(__key));" % \
                                  self.probe_hash_name 
                        key_expr = "%s __key = %s;" % \
                                   (self.expr_type, self.expr)
                program = program.replace("DATA_DECL", decl)
                program = program.replace("KEY_EXPR", key_expr) 
                program = program.replace("FILTER", self.filter or "1") 
                program = program.replace("COLLECT", collect)
                return program

        def attach(self, bpf):
                self.bpf = bpf
                if len(self.library) > 0:
                        if self.is_ret_probe:
                                bpf.attach_uretprobe(name=self.library,
                                                  sym=self.function,
                                                  fn_name=self.probe_func_name,
                                                  pid=self.pid or -1)
                        else:
                                bpf.attach_uprobe(name=self.library,
                                                  sym=self.function,
                                                  fn_name=self.probe_func_name,
                                                  pid=self.pid or -1)
                else:
                        if self.is_ret_probe:
                                bpf.attach_kretprobe(event=self.function,
                                                  fn_name=self.probe_func_name)
                        else:
                                bpf.attach_kprobe(event=self.function,
                                                  fn_name=self.probe_func_name)

        def display(self):
                print(self.raw_spec)
                data = self.bpf.get_table(self.probe_hash_name)
                if self.type == "raw":
                        print("\t%-10s %s" % ("COUNT", "EVENT"))
                        for key, value in sorted(data.items(),
                                                 key=lambda kv: kv[1].value):
                                if self._is_string_probe():
                                        key_str = key.key
                                else:
                                        key_str = str(key.value)
                                print("\t%-10s %s = %s" %
                                      (str(value.value), self.expr, key_str))
                elif self.type == "hist":
                        data.print_log2_hist(val_type=self.expr)

examples = """
Probe specifier syntax:
        <raw|hist>[-ret]:[library]:function(signature)[:type:expr[:filter]]
Where:
        <raw|hist> -- collect raw data or a histogram of values
        ret        -- probe at function exit, only @retval is accessible
        library    -- the library that contains the function
                      (leave empty for kernel functions)
        function   -- the function name to trace
        signature  -- the function's parameters, as in the C header
        type       -- the type of the expression to collect
        expr       -- the expression to collect
        filter     -- a filter that is applied to collected values

EXAMPLES:

gentrace.py "hist::__kmalloc(u64 size):u64:size"
        Print a histogram of allocation sizes passed to kmalloc

gentrace.py -p 1005 "raw:c:malloc(size_t size):size_t:size:size==16"
        Print a raw count of how many times process 1005 called malloc with
        an allocation size of 16 bytes

gentrace.py "raw-ret:c:gets():char*:@retval"
        Snoop on all strings returned by gets()

gentrace.py -p 1005 "raw:c:write(int fd):int:fd"
        Print raw counts of how many times writes were issued to a particular
        file descriptor number, in process 1005

gentrace.py -p 1005 "hist-ret:c:read()"
        Print a histogram of error codes returned by read() in process 1005

gentrace.py "hist:c:write(int fd, const void *buf, size_t count):size_t:count:fd==1"
        Print a histogram of buffer sizes passed to write() across all
        processes, where the file descriptor was 1 (STDOUT)

gentrace.py "raw:c:fork"
        Count fork() calls in libc across all processes
        Can also use funccount.py, which is easier and more flexible 

gentrace.py \\
        "hist:c:sleep(u32 seconds):u32:seconds" \\
        "hist:c:nanosleep(struct timespec { time_t tv_sec; long tv_nsec; } *req):long:req->tv_nsec"
        Print histograms of sleep() and nanosleep() parameter values

gentrace.py -p 2780 -z 120 "raw:c:write(int fd, char* buf, size_t len):char*:buf:fd==1"
        Spy on writes to STDOUT performed by process 2780, up to a string size
        of 120 characters 
"""

parser = argparse.ArgumentParser(description=
        "Trace a function and display a summary of its parameter values.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=examples)
parser.add_argument("-p", "--pid", type=int,
        help="id of the process to trace (optional)")
parser.add_argument("-z", "--string-size", default=80, type=int,
        help="maximum string size to read from char* arguments")
parser.add_argument("-i", "--interval", default=1, type=int,
        help="output interval, in seconds")
parser.add_argument("-c", "--count", type=int,
        help="number of outputs")
parser.add_argument("specifier", nargs="+",
        help="the probe specifiers (see examples below)")
args = parser.parse_args()

specifiers = []
for specifier in args.specifier:
        specifiers.append(Specifier(specifier, args.pid))

bpf_source = "#include <uapi/linux/ptrace.h>\n"
for specifier in specifiers:
        bpf_source += specifier.generate_text(args.string_size)

bpf = BPF(text=bpf_source)

for specifier in specifiers:
        specifier.attach(bpf)

count_so_far = 0
while True:
        try:
                sleep(args.interval)
        except KeyboardInterrupt:
                exit()
        print("[%s]" % strftime("%H:%M:%S"))
        for specifier in specifiers:
                specifier.display()
        count_so_far += 1
        if args.count is not None and count_so_far >= args.count:
                exit()
